### PR TITLE
Type conversion between float/complex dtypes

### PR DESCRIPTION
**Pull request by gautierronan**
This PR is an implementation of the feature request https://github.com/pytorch/pytorch/issues/97888, for the implementation of `torch.dtype.to_complex()` and `torch.dtype.to_float()` methods that convert between float and complex dtypes of the same precision.

Disclaimer: it's the first time I code in C++ so hopefully the code is correct, but I'm not super confident about the PR. Any advice/comment is welcome. It's also my first contribution to a large library, so hopefully I'm not doing anything wrong !

@ezyang 

cc @albanD

### CODE REVIEW

**Comment by ezyang** [comment]
Oops, this was different than what I thought I was asking for in https://github.com/pytorch/pytorch/issues/97888#issuecomment-1489446040 Maybe we can still add it? Let's see:

I originally was responding to:

> Thanks for pointing this function out, I didn't know it existed. One drawback is that it doesn't support float to float or complex to complex conversion (which does nothing, but it could come in handy). And it's lacking documentation in case other users would want to same feature as me.

It seemed to me that you were requesting corresponding_real_dtype to return torch.float if given torch.float. This PR, does not seem to do that.

Do you still want this PR? There might be some extra bikeshedding we will need to do, as the methods you're adding have no analogue in numpy.

**Comment by gautierronan** [comment]
> Oops, this was different than what I thought I was asking for in [#97888 (comment)](https://github.com/pytorch/pytorch/issues/97888#issuecomment-1489446040) Maybe we can still add it? Let's see:
> 
> I originally was responding to:
> 
> > Thanks for pointing this function out, I didn't know it existed. One drawback is that it doesn't support float to float or complex to complex conversion (which does nothing, but it could come in handy). And it's lacking documentation in case other users would want to same feature as me.
> 
> It seemed to me that you were requesting corresponding_real_dtype to return torch.float if given torch.float. This PR, does not seem to do that.
> 
> Do you still want this PR? There might be some extra bikeshedding we will need to do, as the methods you're adding have no analogue in numpy.

Well as you wish ! Now that I personally know that `corresponding_real_dtype` exists, I can use it for my own needs. Indeed I can just extend it to support float->float and complex->complex conversion, which would be very simple. 

The question is rather: do you want the method I was initially proposing (i.e. `torch.dtype.to_real`) for other users? I think it's quite handy, but you might have reasons to not want that.

**Comment by ezyang** [comment]
On reflection, I think I will take this PR too, we just need to change it to `to_real`.

**Comment by gautierronan** [comment]
Alright, let me know if there's anything else I can do.

**Review by ezyang — APPROVED** [review]

**Review by ezyang — CHANGES_REQUESTED** [review]
Actually the mechanism for making a new Python binding may not be correct. More coming later

File: torch/csrc/Dtype.cpp [hunk 1]
```diff
@@ -57,6 +58,28 @@ PyObject* THPDtype_reduce(PyObject* _self, PyObject* noargs) {
...
 
+PyObject* THPDtype_to_real(THPDtype* self, PyObject* noargs) {
+  if (at::isFloatingType(self->scalar_type)) {
+    auto scalar_type = self->scalar_type;
+  } else {
+    auto scalar_type = (at::toRealValueType(self->scalar_type));
+  }
+  std::string primary_name, legacy_name;
+  std::tie(primary_name, legacy_name) = (at::getDtypeNames(scalar_type));
+  return THPDtype_New(scalar_type, primary_name);
```
Comment (ezyang) [inline torch/csrc/Dtype.cpp:69]: Use getTHPDtype instead.

**Review by ezyang — CHANGES_REQUESTED** [review]
how to fix it

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
still doesn't compile

**Comment by gautierronan** [comment]
Unsure how to fix this but I'm still very much interested in the PR. Any help is appreciated !

**Pull request closed without merging**